cpp/src/phonenumbers/phonenumberutil.cc

```diff
@@ -1527,12 +1525,76 @@ void PhoneNumberUtil::FormatNsnWithCarrier(const string& number,
       : metadata.intl_number_format();
   const NumberFormat* formatting_pattern =
       ChooseFormattingPatternForNumber(available_formats, number);
+  NumberFormat new_format;
+  bool consistent_optional_rule = TRUE;
+  bool consistent_fg_rule = TRUE;
   if (!formatting_pattern) {
-    formatted_number->assign(number);
-  } else {
-    FormatNsnUsingPatternWithCarrier(number, *formatting_pattern, number_format,
-                                     carrier_code, formatted_number);
+      // Set the pattern to the possible national patterns, check the number
+      // matches this or really we shouldn't be doing this.
+      const RegExp& national_number_pattern =
+          reg_exps_->regexp_cache_->GetRegExp(
+              StrCat("(",
+                     metadata.general_desc().national_number_pattern(),
+                     ")"));
+      std::cout << "Matching pattern: " << StrCat("(",
+                     metadata.general_desc().national_number_pattern(),
+                     ")") << "\n";
+      if (TestNumberLengthAgainstPattern(national_number_pattern,
+                                         number) != IS_POSSIBLE)
+      {
+        formatted_number->assign(number);
+        std::cout << "Fail out." << "\n";
+        return;
+      }
+      new_format.set_pattern(
+          StrCat("(",
+                 metadata.general_desc().national_number_pattern(),
+                 ")"));
+      new_format.set_format("$1");
+      new_format.add_leading_digits_pattern("[0-9]");
+
+      // Determine national prefix formatting rule. Check that all number
+      // formats agree what the rule should be, if they agree, use what they
+      // agree on. Otherwise assume apply national prefix.
+      if (available_formats.size() != 0) {
+        const NumberFormat first_format = available_formats.Get(0);
+        for (RepeatedPtrField<NumberFormat>::const_iterator
+          it = available_formats.begin(); it != available_formats.end(); ++it) {
+          if (first_format.national_prefix_optional_when_formatting() !=
+              it->national_prefix_optional_when_formatting()) {
+            consistent_optional_rule = FALSE;
+          }
+          if (FormattingRuleHasFirstGroupOnly(
+                first_format.national_prefix_formatting_rule()) !=
+              FormattingRuleHasFirstGroupOnly(
+                it->national_prefix_formatting_rule())) {
+            consistent_fg_rule = FALSE;
+          }
+        }
+
+        if ((consistent_optional_rule) &&
```

Comment:
What if there are zero available formats?  You'll not have a `first_format` to use here.